tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+        check_guest_health($guest);
+    }
+
+    # Remove vGPU device
+    remove_vgpu($vgpu);
+
+    # Disable GPU SR-IOV
+    assert_script_run("/usr/lib/nvidia/sriov-manage -d $gpu_device");
+    record_info("SRIOV disabled", $gpu_device);
+
+    # Upload vgpu related logs
+    upload_virt_logs($log_dir, "logs");
+
+    # Redefine guest from their original configuration files
+    restore_original_guests();
+}
+
+sub is_host_ready_for_vgpu_test {
+    if (is_sle('<15-SP3') or !is_kvm_host) {
+        record_info("Host not supported!", "NVIDIA vGPU is only supported on KVM hosts with SLE15SP3 and greater!", result => 'softfail');
+        return;
+    }
+    #check VT-d is supported in Intel x86_64 machines
+    if (script_run("grep -m 1 Intel /proc/cpuinfo") == 0) {
+        assert_script_run "dmesg | grep -E \"DMAR:.*IOMMU enabled\"";
+    }
+    script_run("lsmod | grep -e vfio -e nvidia -e mdev");
+    return 'yes';
+}
+
+sub get_gpu_device {
+    #NVIDIA AMPERE GPU
+    my $gpu_devices = script_output("lspci | grep NVIDIA | cut -d ' ' -f1");
+    foreach (split("\n", $gpu_devices)) {
+        return $_ if script_run("lspci -v -s $_ | grep 'SR-IOV'") == 0;
+    }
+}
+
+sub install_vgpu_manager_and_reboot {
+    my ${dirver_url} = shift;
+    die "The vGPU driver URL requires to be in 'http://...' format!" unless ${dirver_url} =~ /http:\/\/.*\/(.*\.run)/;
+    my ${driver_file} = $1;
+
+    # Check if vGPU manager has been already loaded
+    if (script_run("nvidia-smi") == 0) {
+        my ${driver_version} = script_output("nvidia-smi -q | grep 'Driver Version' | grep -oE \"[0-9.]+\"");
+        if (${driver_file} =~ /${driver_version}/) {
+            record_info("Warning", "vGPU manager ${driver_version} has already been loaded!", result => 'softfail');
+            return;
+        }
+    }
+    # Install vGPU manager
+    download_script(${driver_file}, script_url => ${dirver_url});
+    zypper_call "in kernel-default-devel gcc";
+    assert_script_run("modprobe vfio_pci_core") if is_sle('>=15-SP5');
+    assert_script_run("./${driver_file} -s");
+    record_info("vGPU manager is installed successfully.");
+
+    # Reboot host(it took 180+ seconds for ph052 to bootup with calltrace)
+    enter_cmd("grub2-once 0; reboot");
+    record_info("Host rebooting ...");
+    select_console 'sol', await_console => 0;
+    # Wait for reboot and show screenshots
+    foreach (1 .. 10) {
+        save_screenshot;
+        sleep 20;
+    }
+    assert_screen([qw(sol-console-wait-typing-ret linux-login text-login)], 120);
+    if (match_has_tag('sol-console-wait-typing-ret')) {
+        send_key 'ret';
+        assert_screen([qw(inux-login text-login)], 120);
+    }
+    record_info("Host rebooted done");
+    reset_consoles;
+    select_console('root-ssh');
+
+    # Verify vGPU manager installed successfully
+    assert_script_run("lspci -d 10de: -k");
+    assert_script_run("lsmod | grep nvidia");
+    assert_script_run("nvidia-smi");
```

Comment:
Better to add `save_screenshot;` under `assert_script_run("nvidia-smi");`

Reply:
`assert_script_run` will save_screenshot automatically.

Reply:
Got it. Make sense now.